File: backend/management/src/main/java/dev/tpcoder/empfriendly/management/point/ExternalPointService.java
package dev.tpcoder.empfriendly.management.point;

import java.math.BigDecimal;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Service
public class ExternalPointService implements PointService {

  private final Logger logger = LoggerFactory.getLogger(ExternalPointService.class);

  private final WebClient webClient;

  public ExternalPointService(@Qualifier("pointWebClient") WebClient webClient) {
    this.webClient = webClient;
  }

  @Override
  public Mono<Void> topup(PointChangeRequest body) {
    logger.debug("Topup request: {}", body);
    if (body.point().compareTo(BigDecimal.ZERO) < 0) {
      throw new RuntimeException("Can't Negative");
    }
    return this.webClient.post()
        .uri("/points/topup")
        .header("requestUid", UUID.randomUUID().toString())
        .bodyValue(body)
        .retrieve()
        .toBodilessEntity()
        .then();
  }

  @Override
  public Mono<Void> deduct(PointChangeRequest body) {
    if (body.point().compareTo(BigDecimal.ZERO) < 0) {
      throw new RuntimeException("Can't Negative");
    }
    logger.debug("Deduct request: {}", body);
    return this.webClient.post()
        .uri("/points/deduct")
        .header("requestUid", UUID.randomUUID().toString())
        .bodyValue(body)
        .retrieve()
        .toBodilessEntity()
        .then();
  }

  @Override
  public Mono<Point> getPointByRefId(String refId) {
    var path = "/points/{referenceId}";
    logger.debug("getPointByRefId refId: {}", refId);
    return this.webClient.get()
        .uri(uriBuilder -> uriBuilder.path(path).build(refId))
        .header("requestUid", UUID.randomUUID().toString())
        .retrieve()
        .bodyToMono(Point.class);
  }
}

File: backend/general/src/main/kotlin/dev/tpcoder/empfriendly/general/point/ExternalPointService.kt
package dev.tpcoder.empfriendly.general.point

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.*

@Service
class ExternalPointService(@Qualifier("pointWebClient") private val pointWebClient: WebClient) :
    PointService {

    private val logger = LoggerFactory.getLogger(ExternalPointService::class.java)

    override fun getTransactionHistory(id: String): Flux<Transaction> {
        val path = "/points/history/$id"
        logger.debug("Get transaction history id: $id")
        return this.pointWebClient.get()
            .uri(path)
            .header("requestUid", UUID.randomUUID().toString())
            .retrieve()
            .bodyToFlux(Transaction::class.java)
    }

    override fun getPointByRefId(referenceId: String): Mono<Point> {
        logger.debug("Get getPointByRefId: $referenceId")
        val path = "/points/$referenceId"
        return this.pointWebClient.get()
            .uri(path)
            .header("requestUid", UUID.randomUUID().toString())
            .retrieve()
            .bodyToMono(Point::class.java)
    }
}